url29ecd5d728_geodb_db3dcore_tgz-pJ8-de_uos_igf_db3d_dbms_model3d_SegmentNetBuilderJ8/src/de/uos/igf/db3d/dbms/geom/GeomUtils.java

```diff
@@ -0,0 +1,1416 @@
+
+				for (int j = 0; j < inactiveBlength; ++j) {
+					kktM[bStart - j][rowCol] = kktM[rowCol][bStart - j] = LA
+							.dot(b[piB[j]], point);
+				}
+				for (int i = 0; i < inactiveAlength; ++i) {
+					kktM[i][rowCol] = kktM[rowCol][i] = -LA.dot(a[piA[i]],
+							point);
+				}
+				kktM[maxInactive + 1][rowCol] = kktM[rowCol][maxInactive + 1] = 1d;
+
+				++inactiveBlength;
+
+				systemHasChanged = true;
+			}
+		}
+
+		return result;
+	}
+
+	/**
+	 * Computes the (square of) the distance between the convex hulls of the two
+	 * given point sets and eventually also the convex coordinates of points at
+	 * minimal distance.
+	 * 
+	 * If a coeffs parameter is specified the corresponding coefficients of the
+	 * convex combination of an accordung closest point is specified.
+	 * {@link http://en.wikipedia.org/wiki/Convex_combination}
+	 * 
+	 * The algorithm solves the optimization problem minimize length(coeffsA[0]
+	 * * a[0] + ... + coeffsA[n] * a[n] - coeffsB[0]*b[0] - ... -
+	 * coeffsB[m]*b[m]) as a function in the coefficients respecting the
+	 * conditions coeffsA[0] + ... + coeffsA[n] = 1 coeffsB[0] + ... +
+	 * coeffsB[m] = 1 coeffsA[i] >= 0 for all i = 0...n coeffsB[j] >= 0 for all
+	 * j = 0...n.
+	 * 
+	 * The linear equation system kktM * kktX = kktY is iteratively constructed
+	 * using the Lagrangians of the previous solution and at each iteration step
+	 * adding the point which is most "inclined" towards the other simplex.
+	 * {@link http://en.wikipedia.org/wiki/Quadratic_programming}
+	 * 
+	 * The algorithm uses the modified version of the active set method
+	 * {@link http://en.wikipedia.org/wiki/Active_set} as described at the Forum
+	 * Bauinformatik 2009 {@link http
+	 * ://digbib.ubka.uni-karlsruhe.de/volltexte/1000012014} (pp. 93-103), by
+	 * the author:
+	 * 
+	 * @param a
+	 *            an array of points defining an affine space clA
+	 * @param b
+	 *            an array of points defining another affine space clB
+	 * @param coeffsA
+	 *            if not null, the affine coefficients for a point pa at minimal
+	 *            distance to clB, i.o.w.
+	 *            <code>pa = sum coeffsA[i]*a[i] for i in 0..a.length</code>,
+	 *            will be stored here. If the specified array is too short a
+	 *            ArrayIndexOutOfBoundsException will occur.
+	 * @param coeffsB
+	 *            if not null, the affine coefficients for a point pb at minimal
+	 *            distance to clA, i.o.w.
+	 *            <code>pb = sum coeffsB[j]*b[j] for j in 0..b.length</code>,
+	 *            will be stored here. If the specified array is too short a
+	 *            ArrayIndexOutOfBoundsException will occur.
+	 * @param scOp
+	 *            the Scalar operator which compares double values up to a
+	 *            specified error.
+	 * @return the square of the distance between the convex closure of a and
+	 *         the convex closure of b.
+	 */
+	public static double simplexDistance(double[][] a, double[][] b,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 50. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.